### PR TITLE
MQTT with QOS 1 or 2 forbids a packet ID of 0.

### PR DESCRIPTION
**Pull request by dlizotte-uwo**
- **Describe the scope of your change--i.e. what the change does and what parts
  of the code were modified.**  This will help us understand any risks of integrating
  the code.

_The change is to the packet numbering, which now starts at 1 and wraps around to 1 (rather than 0) when overflowing 16 bits. This is because the MQTT standard requires nonzero packet IDs for QOS=1 and QOS=2. In the previous version, if a client is created, then connected, then a QOS=1 subscription packet is sent with packet ID 0 is sent, and my server (mosquitto) disconnects. Re-connecting and trying again works because the packet ID has incremented to 1, which is legal._

- **Describe any known limitations with your change.**  For example if the change
  doesn't apply to a supported platform of the library please mention it.

_I don't think there are any._

- **Please run any tests or examples that can exercise your modified code.**  We
  strive to not break users of the code and running tests/examples helps with this
  process.

_I am using this code in a project and have tried with pub/sub; seems to work now._

### CODE REVIEW

File: Adafruit_MQTT.cpp [hunk 1]
```diff
@@ -116,7 +116,7 @@ Adafruit_MQTT::Adafruit_MQTT(const char *server, uint16_t port, const char *cid,
   will_qos = 0;
   will_retain = 0;
 
-  packet_id_counter = 0;
+  packet_id_counter = 1; // MQTT spec forbids packet id of 0 if QOS=1
```
Comment (brentru) [inline Adafruit_MQTT.cpp:119]: @dlizotte-uwo this constructor does not guarantee a QoS level, we should handle checking and incrementing the PID based on the QoS level within the SUBSCRIBE, UNSUBSCRIBE, and PUBLISH calls.

> SUBSCRIBE, UNSUBSCRIBE, and PUBLISH (in cases where QoS > 0) Control Packets MUST contain a non-zero 16-bit Packet Identifier [MQTT-2.3.1-1].

File: Adafruit_MQTT.cpp [hunk 2]
```diff
@@ -137,7 +137,7 @@ Adafruit_MQTT::Adafruit_MQTT(const char *server, uint16_t port,
...
```
Comment (brentru) [inline Adafruit_MQTT.cpp:140]: Same here

File: Adafruit_MQTT.cpp [hunk 3]
```diff
@@ -709,8 +709,8 @@ uint16_t Adafruit_MQTT::publishPacket(uint8_t *packet, const char *topic,
     p[1] = packet_id_counter & 0xFF;
     p += 2;
 
-    // increment the packet id
-    packet_id_counter++;
+    // increment the packet id, skipping 0
+    packet_id_counter = packet_id_counter + 1 + (packet_id_counter + 1 == 0);
```
Comment (brentru) [inline Adafruit_MQTT.cpp:713]: Perform this within the check for QOS<1 on L729 instead

**Review by brentru — CHANGES_REQUESTED** [review]

**Comment by brentru** [comment]
@dlizotte-uwo are you available to make the requested changes above?

**Comment by dlizotte-uwo** [comment]
> @dlizotte-uwo are you available to make the requested changes above?

Hi sorry @brentru - day job is madness. I can probably fix this weekend if that's still useful.

**Comment by brentru** [comment]
@dlizotte-uwo It's still useful! I'll keep this issue open for when you're ready.

File: Adafruit_MQTT.cpp (hunk 1 above)
Comment (fipwmaqzufheoxq92ebc) [reply Adafruit_MQTT.cpp:119]: @brentru Hi, the QoS only needs to be checked for PUBLISH (and is already done in [L707](https://github.com/adafruit/Adafruit_MQTT_Library/blob/7b2061ad85e9ba322048fb53a5e30324bd88334f/Adafruit_MQTT.cpp#L707)). 
SUBSCRIBE and UNSUBSCRIBE always need a (valid) packet id
[(Ref)](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc385349217)
In PUBLISH QoS = 0, the packet id is not used.

File: Adafruit_MQTT.cpp (hunk 3 above)
Comment (fipwmaqzufheoxq92ebc) [reply Adafruit_MQTT.cpp:713]: @brentru This is already in a check [L707](https://github.com/adafruit/Adafruit_MQTT_Library/blob/7b2061ad85e9ba322048fb53a5e30324bd88334f/Adafruit_MQTT.cpp#L707) for QoS > 0 (and for QoS = 0 the counter is not changed and not used)

**Comment by dlizotte-uwo** [comment]
I think based on what [fipwmaqzufheoxq92ebc](https://github.com/fipwmaqzufheoxq92ebc) says and based on https://github.com/adafruit/Adafruit_MQTT_Library/pull/199 that this is still needed and checks are in the right place. I think it's good to keep 1 <= packet_id_counter <= 65535 an invariant in this code.

**Comment by tyeth** [comment]
I'm digesting, I'll come back to you tomorrow.
 I hadn't before appreciated the packet Id of server initiated messages and
client initiated messages could have the same packet Id at the same time.
I'd never really examined the protocol in such depth before. Just verifying
where we check and use packet Id's, etc.

**Comment by tyeth** [comment]
@brentru I think we might be okay, I'm fine with where the checks are, and the use of packet id seems like we wont run into conflicts.

**Comment by brentru** [comment]
Sweet! Thank you for the PR and for your patience, @dlizotte-uwo

**Review by brentru — APPROVED** [review]

**Comment by dlizotte-uwo** [comment]
Thanks @brentru! Sorry for that year I was MIA.

**Pull request merged**